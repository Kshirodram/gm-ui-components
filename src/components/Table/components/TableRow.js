import React from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";
import { transparentize } from "polished";

import { columnItemShape, dataItemShape } from "../types";

import TableCell from "./TableCell";

const TableRowElement = styled.tr.attrs({
  tabIndex: 0
})`
  cursor: pointer;
  position: relative;
  box-shadow: 0 -1px 0 ${props => props.theme.COLOR_KEYLINE};

  /* Give the table row a border when selected */
  ${props =>
    props.isSelected &&
    css`
      background-color: ${props =>
        transparentize(0.85, props.theme.COLOR_BRAND_A)}
      &,
      & + &,
      & + tr {
        box-shadow: none;
      }
      ${props.selectedRowStyle};
    `}
`;

TableRowElement.displayName = "TableRowElement";

function TableRow({
  data,
  columns,
  isSelected,
  onRowClick,
  rowIndex,
  selectedRowStyle
}) {
  return (
    <TableRowElement
      isSelected={isSelected}
      selectedRowStyle={selectedRowStyle}
      onKeyDown={event => {
        if (event.keyCode === 13) {
          const targetCell = event.target.getAttribute("data-column");
          onRowClick({ clicked: data, rowIndex, event, targetCell });
        }
      }}
      onContextMenu={event => {
        const targetCell = event.target.getAttribute("data-column");
        onRowClick({ clicked: data, rowIndex, event, targetCell });
      }}
      onClick={event => {
        const targetCell = event.target.getAttribute("data-column");
        onRowClick({ clicked: data, rowIndex, event, targetCell });
      }}
    >
      {/* Because the `columns` array determines the desired column order, 
        we need to map through it and use the dataIndex property to pick out 
        the appropriate data for that column */}
      {columns.map(({ dataIndex }) => {
        const cellContent = data[dataIndex];
        return (
          <TableCell key={`${dataIndex}|${data.key}`} data-column={dataIndex}>
            {/* Cell content can be either text or a render prop */}
            {typeof cellContent === "function"
              ? cellContent(dataIndex, data, rowIndex)
              : cellContent}
          </TableCell>
        );
      })}
    </TableRowElement>
  );
}

TableRow.propTypes = {
  columns: PropTypes.arrayOf(columnItemShape),
  data: dataItemShape,
  isRowSelected: PropTypes.bool,
  isSelected: PropTypes.bool,
  onRowClick: PropTypes.func,
  rowIndex: PropTypes.number,
  selectedRowStyle: PropTypes.object
};

export default TableRow;
