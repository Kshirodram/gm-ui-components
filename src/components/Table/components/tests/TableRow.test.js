import React from "react";
import { shallow } from "enzyme";

import TableRow from "../TableRow";

let jestSpy = jest.fn();
const columns = [
  {
    title: "Name",
    dataIndex: "name",
    key: "name",
    width: "50%"
  },
  {
    title: "Address",
    dataIndex: "address",
    key: "address",
    width: "20%"
  },
  {
    title: "Age",
    dataIndex: "age",
    key: "age",
    width: "10%"
  },
  {
    title: "Food",
    dataIndex: "favfood",
    key: "favfood",
    width: "10%"
  }
];

const data = {
  key: "2",
  name: "Jane Green",
  age: jestSpy,
  address: "London No. 1 Lake Park",
  favfood: "Amatriciana"
};

describe("TableRow", () => {
  let TableRowWrapper;
  let mockEvent = {
    target: {
      value: "test",
      getAttribute: () => {
        return "data-column-cell-1";
      }
    }
  };

  // expected event result for onclick, onContextMenu, onKeyDown
  const explectedEventResult = {
    clicked: data,
    rowIndex: 1,
    event: mockEvent,
    targetCell: "data-column-cell-1"
  };

  beforeEach(() => {
    TableRowWrapper = shallow(
      <TableRow
        columns={columns}
        data={data}
        onRowClick={jestSpy}
        rowIndex={1}
      />
    );
  });

  test("matches snapshot", () => {
    expect(TableRowWrapper).toMatchSnapshot();
  });

  test("renders a Table Row", () => {
    expect(TableRowWrapper.find("TableRowElement")).toHaveLength(1);
  });

  test("renders the correct amount of cells", () => {
    expect(TableRowWrapper.find("TableCell")).toHaveLength(4);
  });

  test("renders cells in the correct columns", () => {
    TableRowWrapper.find("TableCell").forEach((cell, i) => {
      expect(cell.props()["data-column"]).toBe(columns[i].dataIndex);
    });
  });

  test("invokes a render prop if provided in cell data object", () => {
    expect(jestSpy).toHaveBeenCalled();
  });

  test("on click, should return correct row data", () => {
    TableRowWrapper.simulate("click", mockEvent);

    const lastCall = jestSpy.mock.calls.length - 1;

    expect(jestSpy.mock.calls[lastCall][0]).toEqual(explectedEventResult);
  });

  test("on contextMenu, should return correct row data", () => {
    TableRowWrapper.simulate("contextmenu", mockEvent);

    const lastCall = jestSpy.mock.calls.length - 1;

    expect(jestSpy.mock.calls[lastCall][0]).toEqual(explectedEventResult);
  });

  test("on keydown enter, should return correct row data", () => {
    mockEvent.keyCode = 13;

    TableRowWrapper.simulate("keyDown", mockEvent);

    const lastCall = jestSpy.mock.calls.length - 1;

    expect(jestSpy.mock.calls[lastCall][0]).toEqual(explectedEventResult);
  });
});
