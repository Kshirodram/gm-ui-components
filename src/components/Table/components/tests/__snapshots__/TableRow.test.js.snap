// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow matches snapshot 1`] = `
<TableRowElement
  onClick={[Function]}
  onContextMenu={[Function]}
  onKeyDown={[Function]}
>
  <TableCell
    data-column="name"
    key="name|2"
  >
    Jane Green
  </TableCell>
  <TableCell
    data-column="address"
    key="address|2"
  >
    London No. 1 Lake Park
  </TableCell>
  <TableCell
    data-column="age"
    key="age|2"
  />
  <TableCell
    data-column="favfood"
    key="favfood|2"
  >
    Amatriciana
  </TableCell>
</TableRowElement>
`;
